### PR TITLE
feat[next]: Refactor CSE pass to support ITIR.Program (GTIR branch)

### PR DESCRIPTION
**Pull request by tehrengruber**
Extends the common subexpression elimination to support the new `itir.Program` node and pushes the intermediate `Fencil` -> `Program` conversion upwards the pass manager. The CSE pass now uses the type inference such that only field expressions or composites thereof are collected in field-view context (i.e. outside of `as_fieldop`).

### CODE REVIEW

File: src/gt4py/next/program_processors/codegens/gtfn/gtfn_ir_to_gtfn_im_ir.py [hunk 1]
```diff
@@ -264,25 +264,23 @@ def visit_FunCall(self, node: gtfn_ir.FunCall, **kwargs: Any) -> gtfn_ir_common.
             self.imp_list_ir.append(InitStmt(lhs=gtfn_ir_common.Sym(id=f"{lam_idx}"), rhs=node))
             return gtfn_ir_common.SymRef(id=f"{lam_idx}")
         if isinstance(node.fun, gtfn_ir.Lambda):
```
Comment (tehrengruber) [inline src/gt4py/next/program_processors/codegens/gtfn/gtfn_ir_to_gtfn_im_ir.py:266]: The scoping here was completely broken leading to multiple SymbolName's with the same id being declared in `InitStmt`. The fix is ok, but not great.

File: src/gt4py/next/iterator/ir_utils/ir_makers.py [hunk 2]
```diff
@@ -401,3 +401,15 @@ def _impl(*its: itir.Expr) -> itir.FunCall:
...
+def as_fieldop(expr: itir.Expr, domain: Optional[itir.FunCall] = None) -> call:
+    """
+    Create an `as_fieldop` call.
+
+    Examples
+    --------
+    >>> str(as_fieldop(lambda_("it1", "it2")(plus(deref("it1"), deref("it2"))))("field1", "field2"))
+    '(⇑(λ(it1, it2) → ·it1 + ·it2))(field1, field2)'
+    """
+    return call(call("as_fieldop")(expr, *((domain,) if domain else ())))
```
Comment (egparedes) [inline src/gt4py/next/iterator/ir_utils/ir_makers.py:415]: Maybe it'd be easier to read in this other way...  Feel free to ignore. 
```suggestion
    return call(call("as_fieldop")(*((expr, domain,) if domain else (expr,))))
```

File: src/gt4py/next/iterator/transforms/pass_manager.py [hunk 3]
```diff
@@ -197,12 +203,12 @@ def apply_common_transforms(
     ir = ScanEtaReduction().visit(ir)
 
     if common_subexpression_elimination:
-        ir = CommonSubexpressionElimination().visit(ir)
+        ir = CommonSubexpressionElimination.apply(ir, offset_provider=offset_provider)  # type: ignore[arg-type]
```
Comment (egparedes) [inline src/gt4py/next/iterator/transforms/pass_manager.py:206]: Can you quickly explain the #type ignore comment in the code? I'm not able to guess why is needed

File: src/gt4py/next/iterator/transforms/cse.py [hunk 4]
```diff
@@ -380,23 +386,66 @@ class CommonSubexpressionElimination(PreserveLocationVisitor, NodeTranslator):
...
-            "cartesian_domain",
-            "unstructured_domain",
-        ]:
-            return node
+    @classmethod
+    def apply(
+        cls,
+        node: itir.Program | itir.FencilDefinition | itir.Expr,
+        is_local_view=None,
+        offset_provider=None,
```
Comment (egparedes) [inline src/gt4py/next/iterator/transforms/cse.py:394]: Missing type hints

File: src/gt4py/next/iterator/transforms/cse.py [hunk 5]
```diff
@@ -380,23 +386,66 @@ class CommonSubexpressionElimination(PreserveLocationVisitor, NodeTranslator):
...
-            node, lambda subexpr, num_occurences: num_occurences > 1, self.uids
+        offset_provider = offset_provider or {}
+        node = itir_type_inference.infer(
+            node, offset_provider=offset_provider, allow_undeclared_symbols=not is_program
         )
+        return cls().visit(node, is_local_view=is_local_view)
+
+    def generic_visit(self, node, **kwargs):
+        assert not (cpm.is_call_to("as_fieldop", node) and kwargs.get("is_local_view"))
+        is_local_view = cpm.is_call_to("as_fieldop", node) or kwargs.get("is_local_view")
```
Comment (egparedes) [inline src/gt4py/next/iterator/transforms/cse.py:413]: Just a suggestion:
```suggestion
        parent_is_local_view = kwargs.get("is_local_view")
        is_call_to_as_fieldop = cpm.is_call_to("as_fieldop", node)
        assert not (parent_is_local_view and is_call_to_as_fieldop)
        is_local_view = parent_is_local_view or is_call_to_as_fieldop
```

File: src/gt4py/next/iterator/transforms/cse.py [hunk 6]
```diff
@@ -365,10 +371,10 @@ class CommonSubexpressionElimination(PreserveLocationVisitor, NodeTranslator):
     Perform common subexpression elimination.
```
Comment (egparedes) [inline src/gt4py/next/iterator/transforms/cse.py:372]: Could you briefly outline here the algorithm and the used tools (e.g. other visitors, functions, ...)

File: src/gt4py/next/iterator/transforms/cse.py [hunk 7]
```diff
@@ -380,23 +386,66 @@ class CommonSubexpressionElimination(PreserveLocationVisitor, NodeTranslator):
...
+
+    def visit_FunCall(self, node: itir.FunCall, **kwargs):
+        is_local_view = kwargs["is_local_view"]
+
+        if cpm.is_call_to(node, ("cartesian_domain", "unstructured_domain")):
+            return node
+
+        def predicate(subexpr: itir.Expr, num_occurences: int):
+            # note: be careful here with the syntatic context: the expression might be in local
+            #  view, even though the syntactic context of the root node is in field view.
```
Comment (egparedes) [inline src/gt4py/next/iterator/transforms/cse.py:425]: Just a question: what do you mean by `root` node here? The parent node?

File: src/gt4py/next/iterator/transforms/cse.py [hunk 8]
```diff
@@ -380,23 +386,66 @@ class CommonSubexpressionElimination(PreserveLocationVisitor, NodeTranslator):
...
+                # only extract fields outside of `as_fieldop`
+                # `as_fieldop(...)(field_expr, field_expr)`
+                # -> `(λ(_cs_1) → as_fieldop(...)(_cs_1, _cs_1))(field_expr)`
+                assert isinstance(subexpr.type, ts.TypeSpec)
+                if not is_local_view and all(
+                    isinstance(stype, ts.FieldType)
+                    for stype in type_info.primitive_constituents(subexpr.type)
+                ):
+                    return True
+            return False
```
Comment (egparedes) [inline src/gt4py/next/iterator/transforms/cse.py:439]:
```suggestion
            if num_occurences > 1:
                if is_local_view:
                    return True
                else:
                    # only extract fields outside of `as_fieldop`
                    # `as_fieldop(...)(field_expr, field_expr)`
                    # -> `(λ(_cs_1) → as_fieldop(...)(_cs_1, _cs_1))(field_expr)`
                    assert isinstance(subexpr.type, ts.TypeSpec)
                    return all(
                        isinstance(stype, ts.FieldType)
                        for stype in type_info.primitive_constituents(subexpr.type)
                    )

            return False
```

File: src/gt4py/next/iterator/type_system/type_synthesizer.py [hunk 9]
```diff
@@ -116,15 +116,17 @@ def _(lhs: ts.ScalarType, rhs: ts.ScalarType) -> ts.ScalarType | ts.TupleType:
 
 
 @_register_builtin_type_synthesizer
-def deref(it: it_ts.IteratorType) -> ts.DataType:
+def deref(it: it_ts.IteratorType) -> ts.DataType | ts.DeferredType:
```
Comment (egparedes) [inline src/gt4py/next/iterator/type_system/type_synthesizer.py:119]: Just a question: is the type annotation for `it` correct? Shouldn't it be a union with `DeferredType`? If that is the case, why is not type checker complaining here?
```suggestion
def deref(it: it_ts.IteratorType| ts.DeferredType) -> ts.DataType | ts.DeferredType:
```

File: src/gt4py/next/iterator/type_system/type_synthesizer.py [hunk 10]
```diff
@@ -116,15 +116,17 @@ def _(lhs: ts.ScalarType, rhs: ts.ScalarType) -> ts.ScalarType | ts.TupleType:
...
+        return ts.DeferredType(constraint=None)
     assert isinstance(it, it_ts.IteratorType)
     assert _is_derefable_iterator_type(it)
     return it.element_type
 
 
 @_register_builtin_type_synthesizer
 def can_deref(it: it_ts.IteratorType) -> ts.ScalarType:
-    assert isinstance(it, it_ts.IteratorType)
+    assert isinstance(it, ts.DeferredType) or isinstance(it, it_ts.IteratorType)
```
Comment (egparedes) [inline src/gt4py/next/iterator/type_system/type_synthesizer.py:129]:
```suggestion
    assert isinstance(it, (ts.DeferredType, it_ts.IteratorType))
```

File: src/gt4py/next/iterator/type_system/type_synthesizer.py [hunk 11]
```diff
@@ -324,7 +326,9 @@ def applied_reduce(*args: it_ts.ListType, offset_provider: common.OffsetProvider
 @_register_builtin_type_synthesizer
 def shift(*offset_literals, offset_provider) -> TypeSynthesizer:
     @TypeSynthesizer
-    def apply_shift(it: it_ts.IteratorType) -> it_ts.IteratorType:
+    def apply_shift(it: it_ts.IteratorType) -> it_ts.IteratorType | ts.DeferredType:
+        if isinstance(it, ts.DeferredType):
```
Comment (egparedes) [inline src/gt4py/next/iterator/type_system/type_synthesizer.py:330]: Same question as above

File: src/gt4py/next/program_processors/codegens/gtfn/gtfn_ir_to_gtfn_im_ir.py [hunk 12]
```diff
@@ -264,25 +264,23 @@ def visit_FunCall(self, node: gtfn_ir.FunCall, **kwargs: Any) -> gtfn_ir_common.
...
+            localized_symbols = {**kwargs["localized_symbols"]}  # create a new scope
+            new_kwargs = {**kwargs, "localized_symbols": localized_symbols}
```
Comment (egparedes) [inline src/gt4py/next/program_processors/codegens/gtfn/gtfn_ir_to_gtfn_im_ir.py:268]: I would move the definition of `new_kwargs` to the line before is used, to make it simpler to read.

File: src/gt4py/next/program_processors/codegens/gtfn/gtfn_module.py [hunk 13]
```diff
@@ -181,9 +181,11 @@ def _preprocess_program(
         self,
         program: itir.FencilDefinition,
         offset_provider: dict[str, Connectivity | Dimension],
-    ) -> itir.FencilDefinition | global_tmps.FencilWithTemporaries | itir.Program:
+    ) -> itir.Program:
         if not self.enable_itir_transforms:
-            return program
+            return fencil_to_program.FencilToProgram().apply(
+                program
+            )  # TODO(tehrengruber): should be removed after refactoring to combined IR
```
Comment (egparedes) [inline src/gt4py/next/program_processors/codegens/gtfn/gtfn_module.py:188]: Does this `TODO` imply that it should be tracked in issue #1582 as another cleanup task?

**Review by egparedes — COMMENTED** [review]
It looks good. I have mostly questions and optional suggestions.

**Review by egparedes — APPROVED** [review]
LGTM

**Pull request merged**